lib/twom.c

```diff
@@ -0,0 +1,2510 @@
+    }
+
+    return r;
+}
+
+/* compress 'db', closing at the end.  Uses foreach to copy into a new
+ * database, then rewrites over the old one */
+
+struct copy_rock {
+    struct twom_db *db;
+    struct twom_txn *tid;
+};
+
+static int copy_cb(void *rock,
+                   const char *key, size_t keylen,
+                   const char *data, size_t datalen)
+{
+    struct copy_rock *cr = (struct copy_rock *)rock;
+    int i;
+
+    /* minimal logic from find_loc and stitch knowing that we're
+     * always writing at the end of a file */
+    struct tm_loc *loc = &cr->db->loc;
+    const char *ptr = LOCPTR(loc);
+    uint8_t level = LEVEL(ptr);
+    for (i = 0; i < level; i++)
+        loc->backloc[i] = loc->offset;
+    loc->is_exactmatch = 0;
+    return store_here(cr->db, key, keylen, data, datalen);
+}
+
+static int replay_cb(void *rock,
+                     const char *key, size_t keylen,
+                     const char *data, size_t datalen)
+{
+    struct copy_rock *cr = (struct copy_rock *)rock;
+    return skipwrite(cr->db, key, keylen, data, datalen, 0);
+}
+
+static int tm_rename(struct twom_db *dst, struct twom_db *db)
+{
+    char *copy = strdup(db->fname);
+    const char *dir = dirname(copy);
+    int r = 0;
+    int dirfd = -1;
+    char *tmp;
+
+    if (!db->nosync) {
+    #if defined(O_DIRECTORY)
+        dirfd = open(dir, O_RDONLY|O_DIRECTORY, 0600);
+#else
+        dirfd = open(dir, O_RDONLY, 0600);
+#endif
+        if (dirfd < 0) {
+            db->error("open directory failed",
+                      "filename=<%s> newname=<%s> directory=<%s>",
+                      db->fname, dst->fname, dir);
+            r = TWOM_IOERROR;
+            goto done;
+        }
+    }
+
+    r = rename(dst->fname, db->fname);
+    if (r) goto done;
+
+    if (!db->nosync) {
+        if (fsync(dirfd) < 0) {
+            db->error("fsync directory failed",
+                      "filename=<%s> newname=<%s> directory=<%s>",
+                      db->fname, dst->fname, dir);
+            // but we can't abort now, we've already renamed, so we need to
+            // carry on and update our object
+        }
+    }
+
+    // now that the rename is done, we can move everything over
+    unlock(db, NULL);
+    tmp = db->fname;
+    empty_db(db);
+    *db = *dst;
+    free(db->fname);
+    db->fname = tmp;
+    free(dst);
+
+done:
+    if (dirfd >= 0) close(dirfd);
+    free(copy);
+    return r;
+}
+
+int twom_should_repack(struct twom_db *db, struct twom_txn **txnp)
+{
+    struct twom_txn *txn = *txnp;
+    if (!txn) return 0;
+    struct tm_file *file = txn->file;
+
+    if (!db) return 0;
+
+    struct tm_header *header = &file->header;
+    if (header->dirty_size > MINREWRITE
+        && header->current_size < 4 * header->dirty_size) return 1;
+    return 0;
+}
+
+int twom_repack(struct twom_db *db, struct twom_txn **txnp)
+{
+    char newfname[1024];
+    struct copy_rock cr;
+    int r = twom_begin(db, TWOM_SHARED, txnp);
```

Comment:
Value stored in `r` is never read.

Reply:
This all got changed

Reply:
This all got changed.  The value stored from `myreplay()` in `r` is never read.